wdl-doc/tests/codebase/data_structures/flag_filter.wdl

```diff
@@ -0,0 +1,158 @@
+## Find it [here](https://broadinstitute.github.io/picard/explain-flags.html).
+##
+## ## Example input JSON
+##
+## ```json
+## {
+##    "flags": {
+##        "include_if_all": "0x3",
+##        "exclude_if_any": "0xF04",
+##        "include_if_any": "0x0",
+##        "exclude_if_all": "0x0"
+##    }
+## }
+## ```
+##
+## ### Explanation
+##
+## The above example JSON represents a `FlagFilter` struct
+## being passed to parameter named `flags`.
+## The `include_if_all` field is set to `0x3` which is `3` in decimal.
+## The `exclude_if_any` field is set to `0xF04` which is `3844` in decimal.
+## The `include_if_any` field is set to `0x0` which is `0` in decimal.
+## The `exclude_if_all` field is set to `0x0` which is `0` in decimal.
+##
+## `3` in decimal can be represented as `000000000011` in 12bit binary.
+## This number means that to be included a read must have the 1st and 2nd bits set.
+## Those bits correspond to the `read paired` and `read mapped in proper pair` flags.
+##
+## `3844` in decimal can be represented as `111100000100` in 12bit binary.
+## This number means that to be excluded a read must have **any** of the
+## 3rd, 9th, 10th, 11th, or 12th bits set.
+## We won't go through what all those bits mean here, but you can find
+## the meanings of the bits in the
+## [SAM flag explainer](https://broadinstitute.github.io/picard/explain-flags.html).
+## In short, those are all flags corresponding to the quality of the read
+## and them being `true` may indicate that the read is of low quality and
+## should be excluded.
+
+version 1.1
+
+struct FlagFilter {
+    String include_if_all  # samtools -f
+    String exclude_if_any  # samtools -F
+    String include_if_any  # samtools --rf
+    String exclude_if_all  # samtools -G
+}
+
+task validate_string_is_12bit_oct_dec_or_hex {
+    meta {
+        description: "Validates that a string is a octal, decimal, or hexadecimal number and less than 2^12."
+        help: "Hexadecimal numbers must be prefixed with '0x' and only contain the characters [0-9A-F] to be valid (i.e. [a-f] is not allowed). Octal number must start with '0' and only contain the characters [0-7] to be valid. And decimal numbers must start with a digit between 1-9 and only contain the characters [0-9] to be valid."
+        outputs: {
+            check: "Dummy output to enable caching."
+        }
+    }
+
+    parameter_meta {
+        number: "The number to validate. See task `meta.help` for accepted formats."
+    }
+
+    input {
+        String number
+    }
+
+    command <<<
+        if [[ "~{number}" =~ ^[1-9][0-9]+$ ]]; then
+            # number is in decimal
+            if [ "~{number}" -lt 4096 ]; then
+                >&2 echo "Input number (~{number}) is valid"
+            else
+                >&2 echo "Input number (~{number}) interpreted as decimal"
+                >&2 echo "But number must be less than 4096!"
+                exit 42
+            fi
+        elif [[ "~{number}" =~ ^0[0-7]{0,4}$ ]] \
```

Comment:
Just a thought for the future—it may be worth pulling in the St. Jude Cloud workflows as a submodule here for a more complete picture.